apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/ClientBuilder.cs

```diff
@@ -0,0 +1,322 @@
+﻿/*
+ * Copyright 2018 Google LLC
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+using Google.Apis.Auth.OAuth2;
+using Grpc.Auth;
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Api.Gax.Grpc
+{
+    /// <summary>
+    /// Base class for API-specific builders.
+    /// </summary>
+    /// <typeparam name="TBuilder">The type of this builder, for chaining purposes</typeparam>
+    /// <typeparam name="TClient">The type of client created by this builder.</typeparam>
+    public abstract class ClientBuilder<TBuilder, TClient> where TBuilder : ClientBuilder<TBuilder, TClient>
+    {
+        /// <summary>
+        /// The endpoint to connect to, or null to use the default endpoint.
+        /// </summary>
+        public ServiceEndpoint Endpoint { get; private set; }
+
+        /// <summary>
+        /// The scopes to use, or null to use the default scopes.
+        /// </summary>
+        public IReadOnlyList<string> Scopes { get; private set; }
+
+        /// <summary>
+        /// The channel credentials to use, or null if credentials are being provided in a different way.
+        /// </summary>
+        public ChannelCredentials ChannelCredentials { get; private set; }
+
+        /// <summary>
+        /// The credential file to use, or null if credentials are being provided in a different way.
+        /// </summary>
+        public string CredentialsFile { get; private set; }
```

Comment:
Might want to name this `CredentialsFilePath` of something, to make it clear it's a file-system path. (Assuming that's what it actually is!)

Reply:
Yes, that's what it is. But having made the change, I don't like it :( Two reasons:

- FilePath sounds a bit redundant. It's a filename or a path
- Path sounds like it *might* be a set of places to look, like the PATH environment variable

`File.*` methods all take a parameter called `path`, so `CredentialsPath` might be okay. Or `CredentialsFilename` perhaps. But I don't like the mixture.